# Logística de envíos: ¿Cuándo llega?

## Mentoría DiploDatos 2019 

### Integrantes:

- Chaves, Noelia
- Frau, Johanna

### Mentora:

- Dal Lago, Virginia

### Práctico: Introducción al aprendizaje automático

## Motivación

En la actualidad, cada vez más productos se comercializan a través de una plataforma online. Una de las principales ventajas de este sistema es que el usuario puede recibir el producto en su domicilio en una fecha determinada. Pero, ¿cómo sabemos qué día va a llegar? ¿A partir de qué datos podemos predecir la demora del envío? En este práctico se trabajará con datos de envíos de MercadoLibre, el e-commerce más grande de Latinoamérica, analizando y modelando el problema de logística de envíos para poder responder ¿cuándo llega?

## Descripción del dataset

**Datos:**  El conjunto de datos seleccionado para realizar el práctico corresponde a un muestreo aleatorio no uniforme de 500.000 envíos de MercadoLibre. Estos envíos fueron realizados en Brasil en el período comprendido entre Octubre de 2018 y Abril de 2019 (las fechas originales han sido modificadas y adaptadas a un período de tiempo diferente, conservando el día de la semana y considerando los feriados correspondientes). Mientras que las fechas han sido modificadas, los horarios registrados en el dataset son los originales. Los datos comprenden variables tanto categóricas como numéricas. 

El dataset cuenta con las siguientes columnas:

- **Sender_state:** Estado de Brasil de donde sale el envío.
- **Sender_zipcode:** Código postal (de 5 dígitos) de donde sale el envío.
- **Receiver_state:** Estado de Brasil a donde llega el envío.
- **Receiver_zipcode:** Código postal (de 5 dígitos) a donde llega el envío.
- **Shipment_type:** Método de envío (normal, express, super).
- **Quantity:** Cantidad de productos en un envío.
- **Service:** Servicio del correo con el cual se realizó un envío.
- **Status:** Estado del envío (set: listo para ser enviado, sent: enviado, done: entregado, failed: no entregado, cancelled: cancelado).
- **Date_created:** Fecha de creación del envío.
- **Date_sent:** Fecha y hora en que se realizó el envío (salió del correo).
- **Date_visit:** Fecha y hora en que se entregó el envío al destinatario.
- **Shipment_days:** Días hábiles entre que el envío fue enviado (salió del correo) y que fue entregado.




## Objetivos generales

  

## Objetivos específicos



## Desarrollo del trabajo

In [0]:
import matplotlib.pyplot as plt
import numpy as np
import pandas
import seaborn as sns


from datetime import timedelta
from IPython.display import HTML
from matplotlib.colors import ListedColormap
from scipy import stats
from sklearn import preprocessing
from sklearn import tree
from sklearn.linear_model import LinearRegression, SGDClassifier, LogisticRegression
from sklearn.metrics import confusion_matrix, classification_report, accuracy_score, mean_squared_error
from sklearn.model_selection import train_test_split, GridSearchCV
from sklearn.neighbors import KNeighborsClassifier
from sklearn.preprocessing import MinMaxScaler,PolynomialFeatures

### Lectura y análisis inicial de los datos

In [0]:
url = "https://raw.githubusercontent.com/cnoeliachaves/csv/master/dataset_sample_corrected.csv"

dataset = pandas.read_csv(url, dtype={'sender_zipcode':'int64',
                              'receiver_zipcode':'int64',
                              'quantity':'int64',
                              'service':'int64'},
                       parse_dates=['date_created','date_sent','date_visit'])
dataset.shape


(500000, 12)

In [0]:
dataset.columns

Index(['sender_state', 'sender_zipcode', 'receiver_state', 'receiver_zipcode',
       'shipment_type', 'quantity', 'service', 'status', 'date_created',
       'date_sent', 'date_visit', 'shipment_days'],
      dtype='object')

In [0]:
dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 500000 entries, 0 to 499999
Data columns (total 12 columns):
sender_state        500000 non-null object
sender_zipcode      500000 non-null int64
receiver_state      500000 non-null object
receiver_zipcode    500000 non-null int64
shipment_type       500000 non-null object
quantity            500000 non-null int64
service             500000 non-null int64
status              500000 non-null object
date_created        500000 non-null datetime64[ns]
date_sent           499971 non-null datetime64[ns]
date_visit          499398 non-null datetime64[ns]
shipment_days       499398 non-null float64
dtypes: datetime64[ns](3), float64(1), int64(4), object(4)
memory usage: 45.8+ MB


In [0]:
np.random.seed(1234) #con propositos de reproducibilidad
dataset.sample(10)

sender_state  sender_zipcode  ...          date_visit  shipment_days
459222           SP            4691  ... 2019-03-11 16:04:00           21.0
28619            SP            3017  ... 2019-03-14 06:32:00           15.0
453020           SP           14405  ... 2019-03-13 16:07:00            2.0
16289            SP            4904  ... 2019-03-20 18:05:00            4.0
337995           SP            4266  ... 2019-03-25 11:17:00            6.0
232178           SP           13290  ... 2019-03-07 11:20:37            1.0
220911           SP            6311  ... 2019-03-08 18:59:00            1.0
33028            SP            7803  ... 2019-03-12 16:01:00            2.0
35358            SP            1536  ... 2019-03-19 13:18:00            8.0
47950            SP           12322  ... 2019-03-18 15:06:00            7.0

[10 rows x 12 columns]

### Limpieza de datos

En esta sección retomaremos algunos puntos del práctico anterior con el objetivo de limpiar nuestro dataset.

#### Shipment day anómalos 

En primer lugar, eliminaremos los valores de shipment days negativos.

In [0]:
def clean_shipment_days(dataset):
    """Le saco los valores negativos"""
    shipments_days = dataset['shipment_days']
    mask = (shipments_days >= 0)
    return dataset[mask] 

In [0]:
dataset_clean = clean_shipment_days(dataset)

####Fechas anómalas

En el análisis realizado en el práctico anterior de las 3 variables con información de fechas y horas encontramos que la variable date_created cuenta con valores que caen fuera del intervalo temporal considerado y que todos ellos se encuentran ubicados en el mes de Junio de 2019. Luego de haber realizado una inspección de esta información anómala llegamos a la conclusión que los mismos podrían deberse a un error en el proceso de carga y que en realidad son datos de Enero de 2019. 

Siguiendo esta idea decidimos en este momento subsanar este posible error de carga haciendo que la información de la variable date_created correspondiente al mes de Junio de 2019 sea ahora del mes de Enero de 2019.

In [0]:
date_created_junio = dataset_clean[dataset_clean['date_created'].map(lambda x: x.month) == 6]
date_created_junio

sender_state  sender_zipcode  ...          date_visit  shipment_days
14265            SP            2350  ... 2019-03-01 11:31:00           17.0
18211            SP            7756  ... 2019-03-01 09:08:00           17.0
27700            SP            5034  ... 2019-03-01 11:54:00           17.0
28540            SP            9060  ... 2019-03-01 12:02:00           17.0
33158            SP           16204  ... 2019-03-01 16:43:00           17.0
38127            SP            2460  ... 2019-03-07 14:03:00            0.0
42758            SP            1042  ... 2019-03-01 16:09:00           17.0
43607            SP            6422  ... 2019-03-01 14:30:00           17.0
45731            SP           14405  ... 2019-03-07 04:27:00            0.0
58502            SP           12950  ... 2019-03-01 07:29:00           17.0
60211            SP           14405  ... 2019-03-07 13:17:00            0.0
66197            SP            3064  ... 2019-03-01 07:00:00           17.0
66531            SP            2465  ... 2019-03-07 06:29:00            0.0
68194            SP           14405  ... 2019-03-07 13:17:00            0.0
69523            SP           14405  ... 2019-03-01 13:23:00           17.0
73186            SP            8773  ... 2019-03-01 14:30:00           38.0
76717            SP           12950  ... 2019-03-07 14:55:00            0.0
77370            SP            2066  ... 2019-02-28 23:27:00           16.0
89064            SP            9961  ... 2019-03-07 13:22:00            0.0
89842            SP            6422  ... 2019-03-07 11:42:00            0.0
93661            SP           13920  ... 2019-03-07 11:15:00            0.0
98843            SP            3064  ... 2019-03-07 07:28:00            0.0
110224           SP           13160  ... 2019-03-01 09:53:00           17.0
115505           SP            1536  ... 2019-03-01 08:17:00           17.0
124780           SP            5307  ... 2019-03-01 11:20:00           17.0
128108           SP            3064  ... 2019-03-01 13:44:00           17.0
132631           SP           16204  ... 2019-03-01 18:01:00           17.0
136249           SP           12950  ... 2019-03-01 02:41:00           17.0
150091           SP            2064  ... 2019-03-01 11:32:00           17.0
152487           SP            7411  ... 2019-03-01 10:45:00           17.0
...             ...             ...  ...                 ...            ...
349507           SP            9780  ... 2019-03-01 14:05:00           17.0
352388           SP            1224  ... 2019-03-01 11:02:00           17.0
353172           SP            7411  ... 2019-03-07 08:22:00            0.0
357102           SP           15820  ... 2019-03-01 13:28:00           17.0
357509           SP            9290  ... 2019-03-01 10:04:00           17.0
358087           SP            2950  ... 2019-03-01 10:00:00           38.0
366896           SP           17512  ... 2019-03-01 06:40:00           17.0
402015           SP            9360  ... 2019-03-01 16:04:00           17.0
403246           SP            4186  ... 2019-03-01 08:15:00           17.0
403511           SP            9060  ... 2019-03-01 12:02:00           17.0
406223           SP           13466  ... 2019-03-07 10:10:00            0.0
408164           SP           16204  ... 2019-03-01 04:47:00           17.0
415973           SP            2950  ... 2019-03-07 04:02:00            0.0
417039           SP            9195  ... 2019-03-01 11:12:00           17.0
420117           SP           16204  ... 2019-03-01 18:01:00           17.0
420767           SP            3165  ... 2019-03-01 12:25:00           17.0
434496           SP           14405  ... 2019-03-01 13:28:00           17.0
444764           SP            2950  ... 2019-03-01 07:57:00           17.0
448671           SP            2180  ... 2019-03-01 11:19:00           38.0
449028           SP            9861  ... 2019-03-01 10:52:00           17.0
457368           SP           14095  ... 2019-03-01 10:

In [0]:
date_created_junio = dataset_clean['date_created'].map(lambda x: 
                                                       x+timedelta(days=-151) 
                                                       if x.month==6 else x) 


In [0]:
dataset_clean.loc[:,'date_created'] = date_created_junio

/usr/local/lib/python3.6/dist-packages/pandas/core/indexing.py:543: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  self.obj[item] = s


In [0]:
dataset_clean['date_created'].dt.month.unique()

array([ 3,  2,  1,  4, 12, 10, 11])

In [0]:
dataset_clean.describe(include='datetime64')

date_created            date_sent           date_visit
count                499350               499350               499350
unique                  124                81276                55668
top     2019-03-05 00:00:00  2019-03-11 15:53:00  2019-03-11 22:00:00
freq                  20424                  113                  621
first   2018-10-21 00:00:00  2018-10-23 13:57:00  2019-02-25 05:36:00
last    2019-04-29 00:00:00  2019-03-29 01:46:20  2019-04-02 15:35:00

En el describe de arriba, podemos ver que la última fecha de creación es posterior a la última fecha de data_sent como date_visit, pero consideramos que dichos datos son correctos, ya que no contamos con el completo dataset.

#### Valores nulos

Veamos cuantos valores nulos tenemos

In [0]:
null_values_count = dataset_clean.isnull().sum()
null_values_count[null_values_count > 0]

Series([], dtype: int64)

Podemos ver que no tenemos valores nulos en el dataset.

#### Limpieza de las fechas anomálas

Ahora limpiaremos el dataset sacando todas aquellas fechas que no se siguen en la línea temporal. 

In [0]:
date_created_before_date_sent = dataset_clean.date_created < dataset_clean.date_sent #Filtramos date_created menores que date_sent

date_sent_before_date_visit = dataset_clean.date_sent < dataset_clean.date_visit #Filtramos date_sent menores que date_visit

date_created_before_date_visit = dataset_clean.date_created < dataset_clean.date_visit #Filtramos date_created menores que date_visit


In [0]:
dataset_clean = dataset_clean[date_created_before_date_sent].dropna() # solo fechas de creacion antes de la fecha de envio
dataset_clean = dataset_clean[date_created_before_date_visit].dropna()# solo fechas de creacion antes de la fecha de visita
dataset_clean = dataset_clean[date_sent_before_date_visit].dropna()# solo fechas de envios antes de la fecha de visita
dataset_clean.sample(5)

/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:2: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  
/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:3: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  This is separate from the ipykernel package so we can avoid doing imports until


sender_state  sender_zipcode  ...          date_visit  shipment_days
61209            SP            9520  ... 2019-03-27 15:24:00            6.0
72136            SP            9510  ... 2019-03-15 08:59:00           11.0
192050           SP            9635  ... 2019-03-28 15:29:00            7.0
287022           SP            8586  ... 2019-03-14 06:12:00            2.0
467512           SP            2180  ... 2019-03-29 22:00:00            4.0

[5 rows x 12 columns]

### La importancia de predecir el valor de la variable shipment_days

Uno de los objetivos principales del análisis propuesto en esta mentoría es predecir (dar un número aproximado de días hábiles) el tiempo entre que se despacha del correo hasta que el mismo llega a la casa del comprador. 

Es decir, queremos predecir el feature shipment_days.

Por lo tanto este sería el target de un algoritmo de aprendizaje automático. Si bien este feature originalmente es información de tipo numérica en el práctico anterior utilizamos enconding para convertirla en una variable categórica. Así, esto nos da la posibilidad de trabajar con diferentes modelos de aprendizaje automático utilizando el mismo feature como target pero en sus diferentes versiones.

Frente a este problema consideramos que el estado de Brasil que recibe el envío (receiver_state) es un feature importante a tener en cuenta (mientras más lejos esté de San Pablo probablemente demore más) y probablemente también el código postal de dicho estado (receiver_zipcode). Por otro lado, el tipo de correo (service) también puede ser una variable importante a tener en cuenta porque la calidad de servicio impactará directamente en los días que tarda en llegar el pedido.



In [0]:
dataset_clean.sample(5)

sender_state  sender_zipcode  ...          date_visit  shipment_days
17790            SP           13880  ... 2019-03-01 14:25:00            1.0
63099            SP            3508  ... 2019-03-08 14:58:00            2.0
461551           SP            2120  ... 2019-03-08 09:32:00            9.0
481950           SP           14161  ... 2019-03-25 18:48:00           15.0
126304           SP            1026  ... 2019-03-07 13:14:00            5.0

[5 rows x 12 columns]

In [0]:
dataset_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 496954 entries, 0 to 499999
Data columns (total 12 columns):
sender_state        496954 non-null object
sender_zipcode      496954 non-null int64
receiver_state      496954 non-null object
receiver_zipcode    496954 non-null int64
shipment_type       496954 non-null object
quantity            496954 non-null int64
service             496954 non-null int64
status              496954 non-null object
date_created        496954 non-null datetime64[ns]
date_sent           496954 non-null datetime64[ns]
date_visit          496954 non-null datetime64[ns]
shipment_days       496954 non-null float64
dtypes: datetime64[ns](3), float64(1), int64(4), object(4)
memory usage: 49.3+ MB


Luego de haber limpiado nuestro dataset observamos que contamos con 496954 filas con información, es decir la limpieza solo afectó aproximadamente al 1% de los datos originales.



### Separación en el conjunto de entrenamiento y evaluación 

In [0]:
dataset_clean['date_visit'].groupby(
                          [dataset_clean['date_visit'].dt.month.rename('month'), 
                           dataset_clean['date_visit'].dt.day.rename('day')]
                           ).agg({'count'})

count
month day       
2     25    2021
      26    1617
      27    2661
      28   21658
3     1    20906
      7    31894
      8    38442
      11   35573
      12   36365
      13   25702
      14   25169
      15   24035
      18   31013
      19   24466
      20   25683
      21   23490
      22   22772
      25   28824
      26   23528
      27   24537
      28   22123
      29    3119
4     1     1309
      2       47